ocaml/lambda/simplif.ml

```diff
@@ -220,22 +220,28 @@ let simplify_exits lam =
   *)
 
   let subst = Hashtbl.create 17 in
-  let rec simplif ~try_depth = function
-  | (Lvar _| Lmutvar _ | Lconst _) as l -> l
+  let rec simplif ~layout ~try_depth l =
+    (* layout is the expected layout of the result: [None] if we want to
+       leave it unchanged, [Some layout] if we need to update the layout of
+       the result to [layout]. *)
+    let result_layout ly = Option.value layout ~default:ly in
```

Comment:
I think that it is possible to replace by

```suggestion
    let result_layout ly =
      match layout with
      | None -> ly
      | Some layout ->
          assert(Typeopt.compatible_layout layout ly);
          layout
    in
```

I think that if all layout are correctly generated from the type, given that staticraise is not part of the source language, all the expressions replaced by a staticraise must have the correct type.

Reply:
Unfortunately, I think this will not work, at least without substantial changes to the producers of code, such as https://github.com/ocaml-flambda/flambda-backend/blob/2142e4c1d8e6d24b83d84e3684bcdeb63ed197a2/ocaml/lambda/matching.ml#L3584 .
Here, the `map_return` function keeps the existing layouts, even if the return value gets transformed to a `Lstaticraise`. It is probably possible to change it to fix this problem, but it is unclear if it is worth it compared to the additional complexity in all producers, as well as adding another unspecified-elsewhere invariant (the return layout of an exit must be the bottom layout).